File: java/src/tk/csparpa/gofdp/prototype/ConcretePrototype.java
package tk.csparpa.gofdp.prototype;

import java.util.Date;

/**
 * Concrete implementation of the abstract interface for object cloning.
 */
public class ConcretePrototype extends AbstractPrototype {
	
	private String name;
	private int age;
	private Date birth;
	
	public ConcretePrototype(String name, int age, Date birth){
		this.name = name;
		this.age = age;
		this.birth = birth;
	}

	/**
	 * Implementation of the cloning interface. Notice that we are relying on the
	 * clone operations implemented by the sub-objects of this class 
	 */
	@Override
	public AbstractPrototype clone() throws CloneNotSupportedException {
		System.out.println("Cloning this object");
		AbstractPrototype theClone = 
				new ConcretePrototype(new String(this.name),
									this.age,
									(Date) this.birth.clone());
		return theClone;
	}
	
	@Override
	public void greet(){
		System.out.println("Hi I am a concrete protoype, name="+
				this.name+" age="+Integer.toString(this.age)+
				" born on="+this.birth.toString());
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + age;
		result = prime * result + ((birth == null) ? 0 : birth.hashCode());
		result = prime * result + ((name == null) ? 0 : name.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		ConcretePrototype other = (ConcretePrototype) obj;
		if (age != other.age)
			return false;
		if (birth == null) {
			if (other.birth != null)
				return false;
		} else if (!birth.equals(other.birth))
			return false;
		if (name == null) {
			if (other.name != null)
				return false;
		} else if (!name.equals(other.name))
			return false;
		return true;
	}
}

File: python/prototype/concrete_prototype.py
from abstract_prototype import AbstractPrototype
from copy import deepcopy

class ConcretePrototype(AbstractPrototype):

    """
    Concrete implementation of the abstract interface for object cloning.
    """
    
    _name = None
    _age = None
    _data = None
    
    def __init__(self, name, age, data=None):
        self._name = name
        self._age = age
        self._data = data

    """
    Implementation of the cloning interface. Notice that we are relying on
    the deepcopy module to ensure that all instance variables are properly
    copied (recursively along with their sub-objects)
    """
    def clone(self):
        print("Cloning this object")
        the_clone = ConcretePrototype(deepcopy(self._name),
                                    deepcopy(self._age),
                                    deepcopy(self._data));
        return the_clone
    
    def greet(self):
        print("Hi I am a concrete protoype, name="+self._name+ \
              " age="+str(self._age)+" with data="+str(self._data))
    
    """Redefining the == operator so that equality comparisons can be made"""
    def __eq__(self, other):
        return self._name == self._name and self._age == self._age and \
            self._data == self._data